Handle tablet migration failure while streaming

xemul opened the pull request:
It can happen that a node is lost during tablet migration involving that node. Migration will be stuck, blocking topology state machine. To recover from this, the current procedure is for the admin to execute nodetool removenode or replacing the node. This marks the node as "ignored" and tablet state machine can pick this up and abort the migration.

This PR implements the handling for streaming stage only and adds a test for it. Checking other stages needs more work with failure injection to inject failures into specific barrier.

To handle streaming failure two new stages are introduced -- cleanup_target and revert_migration. The former is to clean the pending replica that could receive some data by the time streaming stopped working, the latter is like end_migration, but doesn't commit the new_replicas into replicas field.

refs: #16527

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6707/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6707/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6707/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7930/)
#### Failed Tests (10/25315):
- [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-source])
- [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-destination])
- [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-source])
- [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-destination])
- [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-source])
- [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration[streaming-destination])
- [topology_custom.test_tablets_migration.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_migration_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets_migration.release.1)
- [topology_custom.test_tablets_migration.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_migration_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets_migration.release.2)
- [topology_custom.test_tablets_migration.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_migration_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets_migration.release.3)
- [test_simple_decommission_node_while_query_info[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_decommission_node_while_query_info_2_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_decommission_node_while_query_info[2])
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider6.cloudius-systems.com

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -880,6 +880,10 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         return _topo_sm._topology.get_excluded_nodes().contains(server_id);
     }
 
+    bool is_ignored(raft::server_id server_id) {
```
Why can't we use is_excluded() instead of the new is_ignored()? What's the semantic difference? get_excluded_nodes() currently just returns ignored_nodes.

xemul replied:
Yes, you're right, they're the same. Will fix in v2

xemul commented:
> Failed Tests (10/25315):
>
>    [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-source%5D)
>    [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-destination%5D)
>    [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-source%5D)
>    [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-destination%5D)
>    [test_node_failure_during_tablet_migration[streaming-source]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_source__3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-source%5D)
>    [test_node_failure_during_tablet_migration[streaming-destination]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6707/testReport/junit/%28root%29/test_tablets_migration/Tests___Unit_Tests___test_node_failure_during_tablet_migration_streaming_destination__3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_failure_during_tablet_migration%5Bstreaming-destination%5D)

Release mode doesn't support fault-injects. Will fix in v2

xemul commented:
v2:
- re-use `is_excluded` helper to check for node being ignored
- marked the new test not tun run in release mode as it uses fault-injection

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6727/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6727/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7951/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6727/)
#### Failed Tests (18/25291):
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [test_tablets_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/%28empty%29/Tests___Unit_Tests___test_tablets_migration_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablets_migration)
- [topology_custom.test_tablets_migration.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6727/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_migration_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets_migration.debug.1)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6734/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6734/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7959/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6734/)
#### Failed Tests (2/25295):
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6734/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6734/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6746/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6746/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7971/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6746/)
#### Failed Tests (4/25294):
- [topology_experimental_raft.test_tablets.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6746/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.6)
- [test_tablet_split](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6746/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_split) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_split)
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6746/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6746/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
#### Build Details:
- Duration: 4 hr 7 min
- Builder: i-0c6df21f7c188ab46 (m5ad.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6754/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6754/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7978/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6754/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider7.cloudius-systems.com

avikivity commented:
Please file/update issues for CI flakiness

avikivity commented on locator/tablets.hh:
```diff
@@ -157,6 +157,7 @@ enum class tablet_transition_stage {
     use_new,
     cleanup,
+    revert_migration,
     end_migration,
```
Since the state machine is now non-linear, please follow up with a mermaid diagram in the appropriate place (I don't think there is one).

xemul replied:
There's no one :( I'll add one as a followup, because even with this pr the diagram would still be incomplete

avikivity commented on locator/tablets.hh:
```diff
@@ -157,6 +157,7 @@ enum class tablet_transition_stage {
     write_both_read_new,
     use_new,
     cleanup,
+    cleanup_target,
```
Perhaps we need more explicit names: cleanup_source, cleanup_target_before_revert.

xemul commented:
upd:
- rebased
- perform more profound check for excluded nodes with the help of get_migration_streaming_info() data

xemul commented on service/topology_coordinator.cc:
```diff
@@ -1014,6 +1014,22 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                            break;
+                        }
+
+                        auto leaving = locator::get_leaving_replica(tmap.get_tablet_info(gid.tablet), trinfo);
```
OK, fixed

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6778/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6778/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8003/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6778/)
#### Build Details:
- Duration: 2 hr 10 min
- Builder: spider1.cloudius-systems.com

xemul commented on service/topology_coordinator.cc:
```diff
@@ -1047,6 +1047,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    if (advance_in_background(gid, tablet_state.cleanup, "cleanup_target", [&] {
+                        locator::tablet_replica dst = trinfo.pending_replica;
+                        if (is_excluded(raft::server_id(dst.host.uuid()))) {
+                            rtlogger.info("Tablet cleanup_target of {} on {} skipped because node is excluded", gid, dst);
```
Fixed to be "Tablet cleanup of {} on {} skipped because node is excluded and doesn't need to revert migration"

xemul commented on service/topology_coordinator.cc:
```diff
@@ -1047,6 +1047,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                            rtlogger.info("Tablet cleanup_target of {} on {} skipped because node is excluded", gid, dst);
+                            return make_ready_future<>();
+                        }
+                        rtlogger.info("Initiating tablet cleanup_target of {} on {}", gid, dst);
```
Fixed to be "Initiating tablet cleanup of {} on {} to revert migration"

xemul commented:
upd:
- rebased
- fixed log messages to look less weird
- changed the way new test validates the "stage" parameter (minor change, it is to ease adding new stages testing in the future)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6802/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6802/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8025/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6802/)
#### Build Details:
- Duration: 2 hr 10 min
- Builder: spider7.cloudius-systems.com

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -991,6 +991,23 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
             };
 
+            auto check_excluded_replicas = [&] {
+                    auto tsi = get_migration_streaming_info(get_token_metadata().get_topology(), tmap.get_tablet_info(gid.tablet), trinfo);
```
I've realized now that the fact that we're using topology from get_token_metadata() here may not be safe. for_each_tablet_transition() is yielding while iterating, which means that the topology version we use here to interpret trinfo may be different than the one from trinfo comes from, and the two may be inconsistent. For example, topology may be missing info for replicas. It could happen that this coordinator was demoted and another one took over while we're iterating here. The scenario is not very likely and any computation here in this case will be dropped because group0 update will fail if the coordinator is demoted, but violation of invariants could maybe cause some bad effects like undefined behavior or abort. To be safe, we should work with the same token metadata snapshot which we use to iterate over tablets.

tgrabiec replied:
On the second thought, this runs under group0 guard, which means that token metadata cannot change locally, so we're safe.
